### PR TITLE
chore: Define macros for most of the "events_processed" internal events

### PR DESCRIPTION
**Pull request by bruceg**
As discussed in Slack. Is there another easy target for DRY?

Signed-off-by: Bruce Guenter <bruce@timber.io>

### CODE REVIEW

**Comment by binarylogic** [comment]
Would this work ok for the changes described in https://github.com/timberio/vector/issues/3517?

**Review by fanatid — APPROVED** [review]

**Comment by binarylogic** [comment]
It might be worth waiting on https://github.com/timberio/vector/pull/3586/files to merge this.

**Comment by bruceg** [comment]
Agreed, I will defer until at least #3586 is done and then rebase this.

File: src/internal_events/mod.rs [hunk 1]
```diff
@@ -1,10 +1,66 @@
 use std::borrow::Cow;
 
+macro_rules! define_events_processed {
+    ($ty:ident, $kind:expr, $name:expr) => {
+        pub(crate) struct $ty;
+
+        impl $crate::internal_events::InternalEvent for $ty {
+            fn emit_logs(&self) {
+                trace!(message = "Processed one event.");
```
Comment (Hoverbear) [inline src/internal_events/mod.rs:9]: Hmmm is this something we want?

Comment (bruceg) [reply src/internal_events/mod.rs:9]: Many (most?) of the pre-macro sources included a trace message like that, so I standardized it in the macro.

Comment (Hoverbear) [reply src/internal_events/mod.rs:9]: Works for me!

**Comment by JeanMertz** [comment]
I worry a bit that this doesn't scale that well —  as seen by the fact that this solution requires a separate/duplicate macro specifically for the ones that also add `bytes_processed`.

As soon as you want to add another counter to either one of these, you'll have to back out of the macro and revert to the manual implementation.

Having said that, I'm okay with merging this, and perhaps we can create an issue to come up with a more generic macro that reduces boilerplate, while still allowing multiple logs/metrics per event?

Another solution (although I'm not sure it's a good fit for our code base) is to do something similar to what we did in Alloy, where the code is auto-generated from a set of YAML files that define the internal events:

* https://github.com/timberio/alloy-api/tree/master/events
* https://github.com/timberio/alloy-api/blob/master/docs/internal/src/architecture/internal-events.md
* https://github.com/timberio/alloy-api/pull/2314

Of course, the investment we make here should be proportional to the scale at which we plan to use these internal events. If it only takes a couple of minutes each time to iterate on the events, and we don't expect a lot of churn, it might not be worth any more effort than what's done in this PR.

**Comment by bruceg** [comment]
> I worry a bit that this doesn't scale that well — as seen by the fact that this solution requires a separate/duplicate macro specifically for the ones that also add `bytes_processed`.
> 
> As soon as you want to add another counter to either one of these, you'll have to back out of the macro and revert to the manual implementation.

I agree with your concern. There are some of the metrics that have exactly this situation. I tried to do something more sophisticated, but Rust macros are a little _too_ hygenic to make this work well.

Another option, in addition to the external tool you reference, would be a procedural macro to derive `InternalEvent` from the struct definition and some additional attributes. This would allow all the code generation to still happen within the normal build process without invoking a separate tool.

> Of course, the investment we make here should be proportional to the scale at which we plan to use these internal events. If it only takes a couple of minutes each time to iterate on the events, and we don't expect a lot of churn, it might not be worth any more effort than what's done in this PR.

Agreed.

**Comment by JeanMertz** [comment]
> Another option, in addition to the external tool you reference, would be a procedural macro to derive `InternalEvent` from the struct definition and some additional attributes. This would allow all the code generation to still happen within the normal build process without invoking a separate tool.

This would be my preference, but I know it's something @lukesteensen also considered when he wrote the RFC, and concluded the work needed to get that up and running wasn't worth the investment. Maybe that balance has changed now, since we're adding more and more internal events.

Another benefit here is that we can (try to) keep naming more consistent than is currently the case. I'm already seeing differences in naming, both for structs, fields names, string formatting (for which you already added a nice awk script to lint against), and whether to add logs such as "One event processed.".

--

I'm happy to 1) merge this PR as an initial ergonomics gain, knowing that it won't scale, and then 2) discuss planning some extra time in the coming weeks to add a procedural macro to derive most of the boilerplate for our internal events and gain some extra consistency as an added bonus.

**Comment by lukesteensen** [comment]
> This would be my preference, but I know it's something @lukesteensen also considered when he wrote the RFC, and concluded the work needed to get that up and running wasn't worth the investment. Maybe that balance has changed now, since we're adding more and more internal events.

I'm still not sure it's worth adding something complex to assist with writing these event structs. They're a little boilerplate-y but quite simple and pretty out of the way. I can create a new one quite easily by copy-pasting an old one and tweaking as needed, and then they're both quite simple to read :smile: 

> Another benefit here is that we can (try to) keep naming more consistent than is currently the case. I'm already seeing differences in naming, both for structs, fields names, string formatting (for which you already added a nice awk script to lint against), and whether to add logs such as "One event processed.".

Consistency in the user-visible output is what I'd consider most important here and hopefully the lints should keep us in line. We could add more structure here, but again I'm just not sure that it's worth adding much complexity.

**Comment by bruceg** [comment]
So is the recommendation then to close this and live with the boilerplate, or to keep just the simplest macro and skip going further?

**Comment by JeanMertz** [comment]
I'm fine merging this, we can iterate later if we want.

Alternatively, we could instead introduce `events_processed!` and `bytes_processed!` macros, they save _a few keystrokes_ over the current `counter!` ones, _and_ they provide a tiny bit of type-safety. e.g. `events_processed!("transform", "add_fields")`.

That way, you can still add other metrics without having to back out of the macro currently introduced in this PR.

Anyway, as I said, happy to see this merge and learn as we go 👍

**Comment by binarylogic** [comment]
@bruce it's your call. If you want to merge this please rebase with `master`. Otherwise, let's close this so we can clean up our PRs.

**Comment by lukesteensen** [comment]
> So is the recommendation then to close this and live with the boilerplate, or to keep just the simplest macro and skip going further?

My intent was to recommend the latter, but I'm happy either way.

**Comment by jszwedko** [comment]
I'm wondering if this is even necessary given the changes in #3888 . We could just use one `EventProcessed` event as it is automatically annotated with the component kind, type, and name.

**Comment by bruceg** [comment]
Given how much the automatic component kind/type/name annotation simplifies these, and that there is no good path to simplifying the rest at the moment, I am fine to just drop this.

**Pull request closed without merging**